lang/java/reef-runtime-azbatch/src/main/java/org/apache/reef/runime/azbatch/driver/AzureBatchResourceManager.java

```diff
@@ -77,22 +80,48 @@ public void onResourceRequested(final ResourceRequestEvent resourceRequestEvent)
   public void onResourceReleased(final ResourceReleaseEvent resourceReleaseEvent) {
     LOG.log(Level.FINEST, "Got ResourceReleasedEvent for Id: {0} in AzureBatchResourceManager",
         resourceReleaseEvent.getIdentifier());
-    this.containerRequests.remove(resourceReleaseEvent.getIdentifier());
+    if (!this.activeContainerIds.remove(resourceReleaseEvent.getIdentifier())) {
+      LOG.log(Level.WARNING,
+          "Attempting to remove non-existent activeContainer for Id: {0} in AzureBatchResourceManager",
+          resourceReleaseEvent.getIdentifier());
+    }
+
+    if (this.containerRequests.containsKey(resourceReleaseEvent.getIdentifier())) {
+      this.containerRequests.remove(resourceReleaseEvent.getIdentifier());
+    } else {
+      LOG.log(Level.WARNING,
+          "Ignoring attempt to remove non-existent container request for Id: {0} in AzureBatchResourceManager",
+          resourceReleaseEvent.getIdentifier());
+    }
     this.evaluatorShimManager.onResourceReleased(resourceReleaseEvent);
   }
 
   public void onResourceLaunched(final ResourceLaunchEvent resourceLaunchEvent) {
+    LOG.log(Level.FINEST, "Got ResourceLaunchEvent for Id: {0} in AzureBatchResourceManager",
+        resourceLaunchEvent.getIdentifier());
     final int evaluatorMemory = this.containerRequests.get(resourceLaunchEvent.getIdentifier()).getMemorySize().get();
     String launchCommand = this.launchCommandBuilder.buildEvaluatorCommand(resourceLaunchEvent,
         evaluatorMemory, this.jvmHeapFactor);
     String evaluatorConfigurationString = this.configurationSerializer.toString(resourceLaunchEvent.getEvaluatorConf());
-
+    this.activeContainerIds.add(resourceLaunchEvent.getIdentifier());
     // TODO[At this point we are only sending the evaluator config and launch command to the shim. AllocatedEvaluator]
     // TODO[object also has "addFile" and "addLibrary" methods, so maybe we'll need to package the JAR for the]
     // TODO[evaluator, too, and not assume that both Evaluator and Shim can share the same JAR.]
     this.evaluatorShimManager.onResourceLaunched(resourceLaunchEvent, launchCommand, evaluatorConfigurationString);
   }
 
+  public Boolean isContainerActive(final String containerId) {
+    return this.activeContainerIds.contains(containerId);
```

Comment:
This searches for an element in a list which is an O(n) operation. Also, it is not thread safe.

Reply:
Used Collections.synchronizedSet on HashSet instead.